pkg/controller/zerocapacity/zero_controller.go

```diff
@@ -0,0 +1,467 @@
+	Init() (*Spec, error)
+	// Perform the operation(s) that are required on the zero-capacity pod
+	Exec(client http.HttpClient) error
+	// Return true when the operation(s) have completed, otherwise false
+	ExecStatus(client http.HttpClient) (Phase, error)
+	// Utility method to return a metav1.Object in order to set the controller reference
+	AsMeta() metav1.Object
+}
+
+type Reconciler interface {
+	// The k8 struct being handled by this controller
+	Type() runtime.Object
+	// Create a new instance of the zero Resource wrapping the actual k8 type
+	ResourceInstance(name types.NamespacedName, ctrl *Controller) (Resource, error)
+}
+
+type Spec struct {
+	// The VolumeSpec to utilise on the zero-capacity pod
+	Volume VolumeSpec
+	// The spec to be used by the zero-capacity pod
+	Container v1.InfinispanContainerSpec
+	// The labels to apply to the zero-capacity pod
+	PodLabels map[string]string
+}
+
+type VolumeSpec struct {
+	// Path within the container at which the volume should be mounted.
+	MountPath string
+	// The VolumeSource to utilise on the zero-capacity pod
+	VolumeSource corev1.VolumeSource
+}
+
+type Controller struct {
+	client.Client
+	Name       string
+	Reconciler Reconciler
+	Kube       *kube.Kubernetes
+	Log        logr.Logger
+	Scheme     *runtime.Scheme
+}
+
+type Phase string
+
+const (
+	// ZeroInitializing means the request has been accepted by the system, but the underlying resources are still
+	// being initialized.
+	ZeroInitializing Phase = "Initializing"
+	// ZeroInitialized means that all required resources have been initialized
+	ZeroInitialized Phase = "Initialized"
+	// ZeroRunning means that the required action has been initiated on the infinispan server.
+	ZeroRunning Phase = "Running"
+	// ZeroSucceeded means that the action on the server has completed and the zero pod has been terminated.
+	ZeroSucceeded Phase = "Succeeded"
+	// ZeroFailed means that the action failed on the infinispan server and the zero pod has terminated.
+	ZeroFailed Phase = "Failed"
+	// ZeroUnknown means that for some reason the state of the action could not be obtained, typically due
+	// to an error in communicating with the underlying zero pod.
+	ZeroUnknown Phase = "Unknown"
+)
+
+func CreateController(name string, reconciler Reconciler, mgr manager.Manager) error {
+	r := &Controller{
+		Name:       name,
+		Client:     mgr.GetClient(),
+		Reconciler: reconciler,
+		Kube:       kube.NewKubernetesFromController(mgr),
+		Log:        logf.Log.WithName(name),
+		Scheme:     mgr.GetScheme(),
+	}
+
+	// Create a new controller
+	c, err := controller.New(name, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource
+	err = c.Watch(&source.Kind{Type: reconciler.Type()}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
```

Comment:
@ryanemerson why are you subscribe for Pod's CRUD events. It can be helpful for intermediate events when Pod's rescheduled by `kubelet` or killed with some errors. I tried to kill Pod during backup by nothing happens.

Reply:
I know why it happens. Operator orienting on the Phase value, not just real situation each reconcile state...